proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,381 @@
+   provided flag assignments) in order to produce a
+   ``PackageDescription``, which is essentially a Cabal file with all
+   conditionals stripped off. Prior to this step, we don't even know
+   what the components of the package necessarily will be (a component
+   can be marked as un-buildable through the dependency solving
+   process).
+
+2. We can now convert a graph of packages into a graph of components. In
+   doing so, the ``ComponentDeps`` tree of dependencies gets exploded into
+   an individual set of package level dependencies. Each component
+   identity is defined by looking at the input dependencies *of the
+   component*, as well other options which would affect the build. Note
+   that in a solver plan, the dependencies refer to *packages*, not
+   components: these package references are resolved to the *public
+   library component* of a package in question. After componentization,
+   ``ComponentDeps`` is eliminated from the install plan.
+
+3. A component is built by configuring its member package for solely the
+   component in question and then building it in a distinct build
+   directory. Execution of the install plan can now be parallelized over
+   components, rather than packages.
+
+**Determining if non-libraries are already installed.** In general, we
+can determine if a library is already installed by consulting the
+installed package database; however, no such database exists for
+executables. However, in ``new-build``, executables are installed to a
+deterministic directory in the Nix store; thus, we can simply check if
+the directory already exists in order to determine if an executable has
+already been built.
+
+**Interaction with packages with legacy Custom scripts.** We can only
+build a package on a per-component basis if the Setup script is
+sufficiently new and supports this interface.  If it does not, we cannot
+do so.  We determine the version of the Setup script in (1) and if
+it is not supported, we treat the package as a legacy node in the
+component graph.  Its component ID is simply the component ID of
+its public library (if it has one; if it does not, no matter, as it's
+not possible to refer to this package as a dependency in any case.)
+
+**Interaction with Backpack.** Backpack needs to perform mix-in linking
+on components, and then a further expansion step to instantiate
+components.  Mix-in linking occurs as we are expanding the package
+graph into a component graph, while instantiation happens as a separate
+step after expansion.
+
+Drawbacks
+---------
+
+* This feature will not be compatible with Custom setup scripts that
+  are linked against an old version of Cabal.
+
+* Package-global ``data-files`` will be duplicated until we consider
+  these files a "component" in-and-of-themselves.
+
+* File paths for installed things will change from where they are
+  currently being stored.
+
+* We need to ``configure`` the package for each component, rather
+  than once for all components.
```

Comment:
This may be mitigated in future at least for non-custom Setup by constructing the configure info directly from cabal-install (minimising the number of expensive and repeated actions).